tests/test_unit/test_shared_type_utils.py

```diff
@@ -0,0 +1,81 @@
+from typing import List, Optional, Union
+
+import pytest
+
+from hera.shared._type_util import (
+    get_annotated_metadata,
+    get_unsubscripted_type,
+    is_annotated,
+    origin_type_issubclass,
+    unwrap_annotation,
+)
+from hera.workflows import Artifact, Parameter
+
+try:
+    from typing import Annotated
+except ImportError:
+    from typing_extensions import Annotated
+
+
+@pytest.mark.parametrize("annotation, expected", [[Annotated[str, "some metadata"], True], [str, False]])
+def test_is_annotated(annotation, expected):
+    assert is_annotated(annotation) == expected
+
+
+@pytest.mark.parametrize(
+    "annotation, expected",
+    [
+        [Annotated[str, Parameter(name="a_str")], str],
+        [Annotated[str, "some metadata"], str],
+        [str, str],
+    ],
+)
+def test_unwrap_annotation(annotation, expected):
+    assert unwrap_annotation(annotation) == expected
+
+
+@pytest.mark.parametrize(
+    "annotation, t, expected",
+    [
+        # Not annotated one.
+        [str, Parameter, None],
+        [Annotated[str, Parameter(name="a_str")], Parameter, Parameter(name="a_str")],
+        [Annotated[str, "some metadata"], Parameter, None],
+        # Must support variadic annotated
+        [Annotated[str, "some metadata", Parameter(name="a_str")], Parameter, Parameter(name="a_str")],
+        # Must support multiple types
+        [Annotated[str, "some metadata", Parameter(name="a_str")], (Parameter, Artifact), Parameter(name="a_str")],
+        # Must consume in order
+        [
+            Annotated[str, "some metadata", Artifact(name="a_str"), Parameter(name="a_str")],
```

Comment:
I'm not a fan of allowing multiple `Parameter`/`Artifact`s in the metadata, we should make this an error (then make the PR a minor version bump). This could be done within `get_annotated_metadata`, or `get_annotated_metadata` should return a list of all metadata matching the type and the caller should check `len(...) == 1`. Option 2 would be a more-correct function implementation (otherwise how do you get the second piece of metadata with the same type?), but it's a private function so... 🤷 Can do option 1 if 2 is too annoying to implement/use

Reply:
@elliotgunton wow great. thanks! I couldn't catch it.

There is some callee that calls this with other types rather than Artifact or Parameter (ModelMapper), so I'll update this as option 2.